OAPE-192: Add verify-requirements CI job for ansible-operator plugins repo

arkadeepsen opened the pull request:
Added verify-requirements CI job for ansible-operator plugins repo. Following steps were used:
- New image `build-requirements` added which uses `Dockerfile.requirements` for creating the requirements files
- New image `get-requirements` added which copies the requirements files from `build-requirements` into `src` which contains the current changes introduced by a PR in ansible-operator-plugins repo
- Added new test `verify-requirements` which uses the `get-requirements` image and runs `git diff --exit-code` to check if there are any changes introduced by the generated requirements files. The test will run if there any changes detected in `images/ansible-operator/Pipfile*`.

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-images

arkadeepsen commented:
`verify-requirements` depends on https://github.com/openshift/ansible-operator-plugins/pull/43

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-verify-collections

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-verify-requirements

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-verify-collections

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-verify-collections

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-verify-collections

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-verify-collections

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-verify-collections

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-verify-collections

arkadeepsen commented:
/pj-rehearse pull-ci-openshift-ansible-operator-plugins-main-verify-requirements

chiragkyal commented on ci-operator/config/openshift/ansible-operator-plugins/openshift-ansible-operator-plugins-main.yaml:
```diff
@@ -22,6 +22,20 @@ images:
 - dockerfile_path: openshift/ci/dockerfiles/ansible-e2e.Dockerfile
   from: openshift-ansible-operator-plugins
   to: ansible-operator-e2e
+- dockerfile_path: openshift/Dockerfile.requirements
+  to: build-requirements
```
I think we should add the following input section to replace the tags and keep it up to date. 
```yaml
inputs:
    ocp_4.19_base-rhel9:
      as:
      - registry.ci.openshift.org/ocp/4.19:base-rhel9
```

chiragkyal commented on ci-operator/config/openshift/ansible-operator-plugins/openshift-ansible-operator-plugins-main.yaml:
```diff
@@ -75,6 +89,12 @@ tests:
+    git diff --exit-code
+  container:
+    from: get-requirements
+  run_if_changed: images/ansible-operator/Pipfile*
```
We should also be watchful for changes done in `openshift/install-ansible.sh` and `openshift/requirements*.txt` files since these are the actual ones that go inside the base image. With this, we'll catch any manual modifications done in these files.

arkadeepsen replied:
The Dockerfile.requirements also needs to be watched as it will impact the generation of the requirements files

chiragkyal commented on ci-operator/config/openshift/ansible-operator-plugins/openshift-ansible-operator-plugins-main.yaml:
```diff
@@ -75,6 +89,12 @@ tests:
     - ref: go-verify-deps
+- as: verify-requirements
+  commands: |
+    git diff --exit-code
```
So, `git` will be able to catch the changes since `get-requirements` image is built from the `src` which would have cloned the repo with the modifications done through `build-requirements`, right?

arkadeepsen replied:
Yes. The pj-rehearse run for the CI job shows the diff for the build requirements files: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/64647/rehearse-64647-pull-ci-openshift-ansible-operator-plugins-main-verify-requirements/1924841604464513024

As there are newer versions available for some of the python packages in the build requirements files these are being used which is causing the diff.

chiragkyal reviewed (commented):
Great work for adding this verification!!. 
I know it wasn't very straightforward, and would have required quite a few test runs.

chiragkyal commented:
/pj-rehearse

arkadeepsen commented:
> Wondering why `verify-collections` is showing up in [#64647 (comment)](https://github.com/openshift/release/pull/64647#issuecomment-2895329603), even if there is no job.

I had added it previously in the same PR, but as it needed additional changes in the ansible-operator-plugins repo I have removed it from this PR.

mytreya-rh commented:
/lgtm

arkadeepsen commented:
/hold cancel
/pj-rehearse ack

mytreya-rh commented:
/lgtm

arkadeepsen commented:
/ph-rehease ack

arkadeepsen commented:
/pj-rehearse ack

chiragkyal commented:
/lgtm
/retest-required

arkadeepsen commented:
/retest
/hold cancel
